types/wlr_virtual_pointer_v1.c

```diff
@@ -0,0 +1,327 @@
+#include <wlr/types/wlr_virtual_pointer_v1.h>
+#include <wlr/types/wlr_pointer.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+#include "wlr-virtual-pointer-unstable-v1-protocol.h"
+
+static void input_device_destroy(struct wlr_input_device *dev) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		(struct wlr_virtual_pointer_v1 *)dev;
+	wl_resource_set_user_data(pointer->resource, NULL);
+	wlr_signal_emit_safe(&pointer->events.destroy, pointer);
+	wl_signal_init(&pointer->events.destroy);
+	wl_list_remove(&pointer->link);
+	free(pointer);
+}
+
+static const struct wlr_input_device_impl input_device_impl = {
+	.destroy = input_device_destroy
+};
+
+static const struct zwlr_virtual_pointer_v1_interface virtual_pointer_impl;
+
+static struct wlr_virtual_pointer_v1 *virtual_pointer_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwlr_virtual_pointer_v1_interface, &virtual_pointer_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void virtual_pointer_motion(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time,
+		wl_fixed_t dx, wl_fixed_t dy) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_motion event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.delta_x = wl_fixed_to_double(dx),
+		.delta_y = wl_fixed_to_double(dy),
+		.unaccel_dx = wl_fixed_to_double(dx),
+		.unaccel_dy = wl_fixed_to_double(dy),
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.motion, &event);
+}
+
+static void virtual_pointer_motion_absolute(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t x, uint32_t y,
+		uint32_t x_extent, uint32_t y_extent) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_motion_absolute event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.x = (double)x / x_extent,
+		.y = (double)y / y_extent,
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.motion_absolute, &event);
+}
+
+static void virtual_pointer_button(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t button,
+		uint32_t state) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_button event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.button = button,
+		.state = state ? WLR_BUTTON_PRESSED : WLR_BUTTON_RELEASED
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.button, &event);
+}
+
+static void virtual_pointer_axis(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t source,
```

Comment:
Ideally, we need to make sure `source` is a valid enum value (note, it's named `axis` in the protocol). Maybe add a protocol error for that.

Reply:
Doesn't libwayland do that for us once we fill out the enum field in the XML?

Reply:
No, it doesn't. That's lame.

Reply:
https://gitlab.freedesktop.org/wayland/wayland/issues/104